### ISSUE TITLE
Always show object `content`, regardless of object `type`, if `mediaType` is supported

### ISSUE THREAD

**Issue by Oblomov** [mastodon/mastodon#24079]
### Pitch

Currently, Mastodon separates ActivityPub objects between “supported” and “converted”, based on object type. Converted objects will only display a title, spoiler text and URL. This is the case even if the object has a `content` that could be handled no differently than the Note type.

A better approach would be to map object properties to the internal model whenever that makes sense, regardless of the object type. This would mean mapping the title & summary to the spoiler text, the (sanitized) content to the text (if supported, i.e. if its  `mediaType` is the default `text/html` or other supported format), etc.

### Motivation

The choice to reduce non-supported types to “title + link” even when content or other compatible attributes are present has no apparent motivation, and leads to issues ranging from the one described in #8892 or #19357 to poor user experience both for Mastodon users and for users of other federate platforms that wish to be more accessible to Mastodon users.

Classic example is the ActivityPub plugin for WordPress recommending to use `Note` instead of `Article` as AP object type, even though `Article` would be the more sensible choice, because otherwise Mastodon followers won't be able to see the post content. Arguably, PixelFed's choice to use `Note` as object type is due to the same reasons.

I know that at least for `Article`s the official rationale behind the choice is that Mastodon is intended for short messages only, but the fact that Mastodon accepts the exact same content when it's presented as `Note` instead of an `Article` (or anything else), the rationale seems ultimately pretty weak, and the _practical_ effect is that other platforms end up abusing the `Note` type to improve the Mastodon users' experience and avoid being relegated to “title and link” on the dominant platform.

So, if there isn't a stronger rationale than “Mastodon is supposed to be for short messages only” (that ship has long sailed), I think it would be better for the entire ecosystem if object attributes were mapped to equivalent status attributes whenever possible, independently from the object type.

**Comment by julianlam** [comment]
This issue and its corresponding PR was referenced in October's ForumWG meeting.

A parallel effort of this by Hometown has already been implemented and is in current use (@dariusk) and I have suggested that he look into your PR to see whether it is a suitable base to build off of (or to merge Hometown's changes upstream).

**Comment by Oblomov** [comment]
Thanks for the update, @julianlam. My PR is just a proof of concept, so I'm sure @dariusk work in Hometown is in a better position for upstreaming. Of course if there's anything I can help with do let me know. (I've also just noticed that my PR has gone stale, I'll try to find some time to rebase it.)

**Comment by LuisAPI** [comment]
Agreed. In many cases, there is no practical difference between `Note` and `Article`, and there should be zero issues with Mastodon attempting to load an Article as items of this object type are similarly mostly text.

**Comment by dahlia** [comment]
Is the team still interested in this? If the direction described here is one you'd be open to, I'd be happy to put together a patch myself.

**Comment by dahlia** [comment]
One thing I'd like to clarify on scope: does this cover types outside the currently converted set too (`Proposal`, or any other type not in `Article`/`Page`/`Image`/`Audio`/`Video`/`Event`), or only upgrading the converted set to show `content`? The title reads as the former, but the pitch text only gives examples from the latter.

**Comment by silverpill** [comment]
I've been using the proposed approach ("duck typing") for a long time, it works very well and improves interoperability. But you may need to prevent processing of certain objects as posts. For example, actors, activities and collections, which also can have the `content` property.

An algorithm for determining the class of an object is described in https://codeberg.org/fediverse/fep/src/branch/main/fep/2277/fep-2277.md

**Comment by scammo** [comment]
From my perspective, this is a really interesting topic that Mastodon should prioritise. This would ensure future compatibility with other Fediverse services and further increase Mastodon's role as a catalyst for the Fediverse as a whole.
If anyone needs any support or feedback, I'm happy to help.

**Comment by julianlam** [comment]
Mastodon in the past year has already improved its handling of "non-Note" object types. Certainly not to a degree I'd like, but — more importantly — it's not Mastodon's fault for not accepting every object type (including all future object types) under the sun.

Its new baseline does show `content`, and with improvements outlined in FEP b2b8, this is no longer really an issue and could potentially be closed @Oblomov ?

**Comment by Oblomov** [comment]
Hi @julianlam (and I think this would also answer @dahlia), my understanding is that tha the work done so far (and FEP b2b8) is still mostly focused only on long-form text (`Article`), not other object types (e.g. `Image` / `Audio` / `Video`). I would prefer if this issue remained open until a broader support is included, possibly following the duck-typing idea presented by @silverpill, even if there there will always be object types that won't be fully supported (especially future types, with possibly currently unknown properties that may need special handling).
